### PR TITLE
1045 repo 4.0

### LINKED ISSUE #1045: Support repo 4.0 API
The 4.0 repo API supports nice things such as authentication. We should target the 4.0 API when possible so users can configure their credentials and access their private packages.

### PR DESCRIPTION
**Pull request by alnlarsen**
Use repo client and target repo 4.0 API to support authentication

Closes #1045

### CODE REVIEW

File: Engine/Authentication/AuthenticationSettings.cs [hunk 1]
```diff
@@ -101,6 +101,7 @@ void PrepareRequest(HttpRequestMessage request, string domain, CancellationToken
                     if (token.RefreshToken != null)
                     {
                         // TODO: refresh
+                        // TODO: fix todo
```
Comment (StefanHolst) [inline Engine/Authentication/AuthenticationSettings.cs:104]: Seems unhelpful

File: Package/Repositories/HttpPackageRepository.cs [hunk 2]
```diff
@@ -34,9 +35,8 @@ public class HttpPackageRepository : IPackageRepository, IPackageDownloadProgres
         private static TraceSource log = Log.CreateSource("HttpPackageRepository");
         private const string ApiVersion = "3.0";
         private VersionSpecifier MinRepoVersion = new VersionSpecifier(3, 0, 0, "", "", VersionMatchBehavior.AnyPrerelease | VersionMatchBehavior.Compatible);
```
Comment (StefanHolst) [inline Package/Repositories/HttpPackageRepository.cs:37]: Is this needed anymore?

File: Package/IPackageRepository.cs [hunk 3]
```diff
@@ -163,7 +163,7 @@ static void ParallelTryForEach<TSource>(IEnumerable<TSource> source, Action<TSou
             var version = id.Version == VersionSpecifier.AnyRelease ? "" : id.Version.ToString();
             string query =
                 @"query Query {
-                            packages(distinctName:true" +
+                            packages(directory: ""/Packages/"", distinctName:true" +
```
Comment (StefanHolst) [inline Package/IPackageRepository.cs:166]: Should we just use the query api from the client?

File: Package/Image/PackageDependencyQuery.cs [hunk 4]
```diff
@@ -17,7 +18,7 @@ static class PackageDependencyQuery
     {
         static string GraphQueryPackages(string os, CpuArchitecture arch, string version, string name) =>
             @"query Query { 
-                packages(version: ""__VERSION__"", type:""tappackage"", os:""__OS__"", architecture:""__ARCH__"" __NAME_QUERY__) {
+                packages(directory: ""/Packages/"", version: ""__VERSION__"", type:""tappackage"", os:""__OS__"", architecture:""__ARCH__"" __NAME_QUERY__) {
```
Comment (StefanHolst) [inline Package/Image/PackageDependencyQuery.cs:21]: Should we use the repoclient here as well?

File: Package/Repositories/HttpPackageRepository.cs [hunk 5]
```diff
@@ -439,6 +321,25 @@ public void DownloadPackage(IPackageIdentifier package, string destination, Canc
             }
         }
 
+        private void Auth()
```
Comment (StefanHolst) [inline Package/Repositories/HttpPackageRepository.cs:324]: Should we call it something like, AttachAuthentication?

File: Package/Repositories/HttpPackageRepository.cs [hunk 6]
```diff
@@ -497,44 +370,15 @@ public string[] GetPackageNames(CancellationToken cancellationToken, params IPac
...
-            else
+            var parameters = new Dictionary<string, object>()
             {
-                using (Stream stream = new MemoryStream())
-                {
-                    compatibleWith = CheckCompatibleWith(compatibleWith);
-                    PackageDef.SaveManyTo(stream, ConvertToPackageDef(compatibleWith));
-                    stream.Seek(0, 0);
-                    string data = new StreamReader(stream).ReadToEnd();
+                ["type"] = @class,
```
Comment (StefanHolst) [inline Package/Repositories/HttpPackageRepository.cs:375]: Type and class are not the same. It should just be class

File: Package/Repositories/HttpPackageRepository.cs [hunk 7]
```diff
@@ -695,8 +548,8 @@ public JObject Query(string query)
         /// </summary>
         /// <param name="query">A GraphQL query string</param>
         /// <returns>A JSON string containing the GraphQL response</returns>
-        public string QueryGraphQL(string query) =>
-            downloadPackagesString($"/3.1/query", query, "application/json", "application/json");
+        public string QueryGraphQL(string query) => downloadPackagesString($"/3.1/query", query, "application/json", "application/json");
```
Comment (StefanHolst) [inline Package/Repositories/HttpPackageRepository.cs:551]: Could we add obsolete attribute and remove this in 9.22? If so, could we also do this for the method above

**Review by StefanHolst — CHANGES_REQUESTED** [review]
Looks great, just some minor comments.

**Review by StefanHolst — APPROVED** [review]
Looks good

**Review by AsgerIversen — APPROVED** [review]

**Pull request merged**